### Positional Cross Matching
***
__Bright Source Sample of AT20G Survey Catalogue__ and the __superCOSMOS all-sky Catalogue__
are the sample datasets used in the following simulation of the positional cross matching
the BSS catalogue is a radio wavlength sky-survey, and superCOMOS is an optical wavelegth sky survey.

### Reading Data

In [1]:
import numpy as np 
import pandas as pd 

In [4]:
bsscat = np.loadtxt('bss.dat',usecols=range(1,7))
bsscat[0]

array([  0.  ,   4.  ,  35.65, -47.  ,  36.  ,  19.1 ])

### Fixed-width columns and loadtxt
loadtxt does not work for fixed-width columns if values are missing. Since there are no missing ID, RA and dec values it is fine for loading the first few columns of the BSS catalogue.

In [6]:
cosmoscat = np.loadtxt('super.csv',delimiter=',',skiprows=1,usecols=[0,1])
cosmoscat[:10]

array([[ 6.03176000e-02, -8.56561327e+01],
       [ 1.14850820e+00, -4.76054131e+01],
       [ 1.27943310e+00, -1.54590140e+00],
       [ 2.64883140e+00, -3.04631581e+01],
       [ 2.75505950e+00, -2.62091826e+01],
       [ 3.24955190e+00, -3.99072049e+01],
       [ 3.64358700e+00, -2.00088103e+01],
       [ 6.02786440e+00, -6.83485009e+01],
       [ 9.27002630e+00, -1.15987340e+00],
       [ 9.55828190e+00, -6.12046403e+01]])

### Normalizing RA and Dec into degrees from the two datasets. 

### AT20G BSS and SuperCOSMOS catalogues from the files bss.dat and super.csv as described 

Each function should return a list of tuples containing the object's ID (an integer) and the coordinates in degrees. The object ID should be the row of the object in the catalogue, starting at 1.

In [7]:
# Write your import_bss function here.
import numpy as np
def hms2dec(hr,mn,sec):
    return (15*(hr + mn/60 + sec/(60*60)))

def dms2dec(deg,arcmin,arcsec):
    dec_deg = abs(deg) + arcmin/60 + arcsec/(60*60)
    
    if deg > 0:
        return dec_deg
    else:  
        return -dec_deg

def import_bss(filename = 'bss.dat'):
    tuple_res = []
    bsscat = np.loadtxt(filename,usecols=range(1,7))
    for idval, row in enumerate(bsscat,1):
        ascension = hms2dec(row[0] , row[1] , row[2])
        declination = dms2dec(row[3],row[4],row[5])
        tuple_res.append((idval,ascension,declination))
    return tuple_res

def import_super(filename='super.csv'):
    tuple_res = []
    supercat = np.loadtxt(filename,delimiter=',',usecols=[0,1],skiprows=1)
    for id_val, row in enumerate(supercat, 1):
        ascension = row[0]
        declination = row[1]
        tuple_res.append((id_val,ascension,declination))
    
    return tuple_res


if __name__ == '__main__':
    bss_cat = import_bss()
    super_cat = import_super()
    print(bss_cat)
    print(super_cat)

[(1, 1.1485416666666666, -47.60530555555556), (2, 2.6496666666666666, -30.463416666666667), (3, 2.7552916666666665, -26.209194444444442), (4, 3.2495416666666666, -39.907333333333334), (5, 6.4549166666666675, -26.03686111111111), (6, 6.568333333333333, -35.21372222222222), (7, 9.561333333333334, -24.98386111111111), (8, 12.497833333333332, -57.641), (9, 12.789583333333333, -42.442361111111104), (10, 14.694333333333335, -56.9865), (11, 15.562791666666666, -80.2111388888889), (12, 15.577708333333334, -75.78138888888888), (13, 16.687958333333334, -40.57208333333334), (14, 19.453374999999998, -21.185388888888887), (15, 19.73875, -21.691694444444444), (16, 20.132125, -27.0235), (17, 21.239041666666665, -51.22113888888889), (18, 23.181374999999996, -16.91338888888889), (19, 23.27404166666667, -52.000972222222224), (20, 23.490000000000002, -36.493027777777776), (21, 23.633916666666668, -38.72602777777778), (22, 24.409708333333334, -24.51488888888889), (23, 25.792208333333335, -32.0154722222222

### find_closest function that takes a catalogue and the position of a target source (a right ascension and declination) and finds the closest match for the target source in the catalogue.

function will return the ID of the closest object and the distance to that object.

The right ascension and declination are in degrees. The catalogue list has been loaded by import_bss from the previous question. The full 320 object BSS catalogue is contained in bss.dat for you to test your code on.

In [8]:
# Write your angular_dist function here.
import numpy as np 
def angular_dist(r1,d1,r2,d2):
    r1,d1,r2,d2 = map(np.radians, [r1,d1,r2,d2])
    a = np.sin(abs(d1-d2)/2)**2
    b = np.cos(d1)*np.cos(d2)*np.sin(abs(r1-r2)/2)**2
    d = 2*np.arcsin(np.sqrt(a+b))
    dec_deg = np.degrees(d)
    return dec_deg

In [9]:
# Write your find_closest function here
import numpy as np 
def find_closest(cat,RA,DEC):
    min_dist = np.inf
    min_id = 0
    for id_val, ra ,dec in cat:
        angular_distance = angular_dist(RA,DEC,ra,dec)
        if angular_distance < min_dist:
            closest = (id_val,angular_distance)
    return closest 


if __name__ == '__main__':
    cat = import_bss()
      
    # First example RA and DEC
    print(find_closest(cat, 175.3, -32.5))

    # Second example RA and DEC
    print(find_closest(cat, 32.2, 40.7))


(160, 16.879012282951326)
(160, 146.42896658367314)


### Final Script

In [10]:
# Write your find_closest function here
import numpy as np 

def hms2dec(hr,mn,sec):
    return (15*(hr + mn/60 + sec/(60*60)))

def dms2dec(deg,arcmin,arcsec):
    dec_deg = abs(deg) + arcmin/60 + arcsec/(60*60)
    return dec_deg if deg > 0 else -dec_deg

def import_bss(filename = 'bss.dat'):
    tuple_res = []
    bsscat = np.loadtxt(filename,usecols=range(1,7))
    for idval, row in enumerate(bsscat,1):
        ascension = hms2dec(row[0] , row[1] , row[2])
        declination = dms2dec(row[3],row[4],row[5])
        tuple_res.append((idval,ascension,declination))
    return tuple_res

def import_super(filename='super.csv'):
    tuple_res = []
    supercat = np.loadtxt(filename,delimiter=',',usecols=[0,1],skiprows=1)
    for id_val, row in enumerate(supercat, 1):
        ascension = row[0]
        declination = row[1]
        tuple_res.append((id_val,ascension,declination))
    
    return tuple_res

def angular_dist(r1,d1,r2,d2):
    r1,d1,r2,d2 = map(np.radians, [r1,d1,r2,d2])
    a = np.sin(abs(d1-d2)/2)**2
    b = np.cos(d1)*np.cos(d2)*np.sin(abs(r1-r2)/2)**2
    d = 2*np.arcsin(np.sqrt(a+b))
    dec_deg = np.degrees(d)
    return dec_deg


def find_closest(cat,RA,DEC):
    min_dist = np.inf
    min_id = None
    closest = (min_id, min_dist)
    for datapoint in cat:
        angular_distance = angular_dist(RA,DEC,datapoint[1],datapoint[2])
        if angular_distance < closest[1]:
            closest = (datapoint[0],angular_distance)
    return closest 


# You can use this to test your function.
# Any code inside this `if` statement will be ignored by the automarker.
if __name__ == '__main__':
  cat = import_bss()
  
  # First example from the question
  print(find_closest(cat, 175.3, -32.5))

  # Second example in the question
  print(find_closest(cat, 32.2, 40.7))


(156, 3.7670580226469053)
(26, 57.729135775621295)


### Crossmatching the BSS and SuperCOSMOS catalogues together to see how many of the bright radio sources in the BSS catalogue have a counterpart in the SuperCOSMOS catalogue

In [11]:
# Write your crossmatch function here.
import numpy as np 

def hms2dec(hr,mn,sec):
    return (15*(hr + mn/60 + sec/(60*60)))

def dms2dec(deg,arcmin,arcsec):
    dec_deg = abs(deg) + arcmin/60 + arcsec/(60*60)
    return dec_deg if deg > 0 else -dec_deg

def import_bss(filename = 'bss.dat'):
    tuple_res = []
    bsscat = np.loadtxt(filename,usecols=range(1,7))
    for idval, row in enumerate(bsscat,1):
        ascension = hms2dec(row[0] , row[1] , row[2])
        declination = dms2dec(row[3],row[4],row[5])
        tuple_res.append((idval,ascension,declination))
    return tuple_res

def import_super(filename='super.csv'):
    tuple_res = []
    supercat = np.loadtxt(filename,delimiter=',',usecols=[0,1],skiprows=1)
    for id_val, row in enumerate(supercat, 1):
        ascension = row[0]
        declination = row[1]
        tuple_res.append((id_val,ascension,declination))
    
    return tuple_res

def angular_dist(r1,d1,r2,d2):
    r1,d1,r2,d2 = map(np.radians, [r1,d1,r2,d2])
    a = np.sin(abs(d1-d2)/2)**2
    b = np.cos(d1)*np.cos(d2)*np.sin(abs(r1-r2)/2)**2
    d = 2*np.arcsin(np.sqrt(a+b))
    dec_deg = np.degrees(d)
    return dec_deg


def find_closest(cat,RA,DEC):
    min_dist = np.inf
    min_id = None
    closest = (min_id, min_dist)
    for datapoint in cat:
        angular_distance = angular_dist(RA,DEC,datapoint[1],datapoint[2])
        if angular_distance < closest[1]:
            closest = (datapoint[0],angular_distance)
    return closest 

def crossmatch(catalogue1,catalogue2,max_dist):
    no_matches = list()
    matches = list()
    
    
    for datapoint in catalogue1:
        closestpoint = find_closest(catalogue2,datapoint[1],datapoint[2])
        if(closestpoint[1] > max_dist):
            no_matches.append(datapoint[0])
        else:
            matches.append((datapoint[0],closestpoint[0],closestpoint[1]))
        
    return matches, no_matches




# You can use this to test your function.
# Any code inside this `if` statement will be ignored by the automarker.
if __name__ == '__main__':
  bss_cat = import_bss()
  super_cat = import_super()

  # First example in the question
  max_dist = 40/3600
  matches, no_matches = crossmatch(bss_cat, super_cat, max_dist)
  print(matches[:3])
  print(no_matches[:3])
  print(len(no_matches))

  # Second example in the question
  max_dist = 5/3600
  matches, no_matches = crossmatch(bss_cat, super_cat, max_dist)
  print(matches[:3])
  print(no_matches[:3])
  print(len(no_matches))


[(1, 2, 0.00010988610938710059), (2, 4, 0.0007649845967242494), (3, 5, 0.00020863352870707666)]
[5, 6, 11]
9
[(1, 2, 0.00010988610938710059), (2, 4, 0.0007649845967242494), (3, 5, 0.00020863352870707666)]
[5, 6, 11]
40


### Calculating the Time Complexity of this Manual Cross Matching

In [12]:
# Write your crossmatch function here.
import numpy as np 
import time 

def hms2dec(hr,mn,sec):
    return (15*(hr + mn/60 + sec/(60*60)))

def dms2dec(deg,arcmin,arcsec):
    dec_deg = abs(deg) + arcmin/60 + arcsec/(60*60)
    return dec_deg if deg > 0 else -dec_deg

def import_bss(filename = 'bss.dat'):
    tuple_res = []
    bsscat = np.loadtxt(filename,usecols=range(1,7))
    for idval, row in enumerate(bsscat,1):
        ascension = hms2dec(row[0] , row[1] , row[2])
        declination = dms2dec(row[3],row[4],row[5])
        tuple_res.append((idval,ascension,declination))
    return tuple_res

def import_super(filename='super.csv'):
    tuple_res = []
    supercat = np.loadtxt(filename,delimiter=',',usecols=[0,1],skiprows=1)
    for id_val, row in enumerate(supercat, 1):
        ascension = row[0]
        declination = row[1]
        tuple_res.append((id_val,ascension,declination))
    
    return tuple_res

def angular_dist(r1,d1,r2,d2):
    r1,d1,r2,d2 = map(np.radians, [r1,d1,r2,d2])
    a = np.sin(abs(d1-d2)/2)**2
    b = np.cos(d1)*np.cos(d2)*np.sin(abs(r1-r2)/2)**2
    d = 2*np.arcsin(np.sqrt(a+b))
    dec_deg = np.degrees(d)
    return dec_deg


def find_closest(cat,RA,DEC):
    min_dist = np.inf
    min_id = None
    closest = (min_id, min_dist)
    for datapoint in cat:
        angular_distance = angular_dist(RA,DEC,datapoint[1],datapoint[2])
        if angular_distance < closest[1]:
            closest = (datapoint[0],angular_distance)
    return closest 

def crossmatch(catalogue1,catalogue2,max_dist):
    no_matches = list()
    matches = list()
    start = time.perf_counter()
    for datapoint in catalogue1:
        closestpoint = find_closest(catalogue2,datapoint[1],datapoint[2])
        if(closestpoint[1] > max_dist):
            no_matches.append(datapoint[0])
        else:
            matches.append((datapoint[0],closestpoint[0],closestpoint[1]))
    seconds = time.perf_counter() - start 
    print('Time Complexity : {}'.format(seconds))
    return matches, no_matches






if __name__ == '__main__':
  bss_cat = import_bss()
  super_cat = import_super()

  # First example 
  max_dist = 40/3600
  matches, no_matches = crossmatch(bss_cat, super_cat, max_dist)
  print(matches[:3])
  print(no_matches[:3])
  print(len(no_matches))

  # Second example 
  max_dist = 5/3600
  matches, no_matches = crossmatch(bss_cat, super_cat, max_dist)
  print(matches[:3])
  print(no_matches[:3])
  print(len(no_matches))


Time Complexity : 1.3306885839999723
[(1, 2, 0.00010988610938710059), (2, 4, 0.0007649845967242494), (3, 5, 0.00020863352870707666)]
[5, 6, 11]
9
Time Complexity : 1.3411538279999604
[(1, 2, 0.00010988610938710059), (2, 4, 0.0007649845967242494), (3, 5, 0.00020863352870707666)]
[5, 6, 11]
40


### Explanation
***
This workflow is the actual logic behind crossmatching technique of two sky-survey catalogues, cross matching is important to verify and to study the object of interest in
different wavelegths
The above work flow took 1.3 seconds to elapse, the sample above consist only a fraction of the actual survey catalogue so it will take forever to elapse on the real catalogue. the next notebook is to improve the above manual process by efficiently using the __Astropy__ library of python which has a functionallity to cross match two catalogues
The library function runs extremly efficient on time complexity since they are built on 
Low lying languages like _C-language_ and _fotran_